lib/ecto/type.ex

```diff
@@ -600,117 +598,95 @@ defmodule Ecto.Type do
 
   """
   @spec cast(t, term) :: {:ok, term} | :error
-  def cast({:embed, type}, value) do
-    cast_embed(type, value)
-  end
-
+  def cast({:embed, type}, value), do: cast_embed(type, value)
   def cast({:in, _type}, nil), do: :error
   def cast(_type, nil), do: {:ok, nil}
 
-  def cast(:binary_id, value) when is_binary(value) do
-    {:ok, value}
+  def cast(type, value) do
+    caster(type).(value)
+  end
+
+  defp caster(type) when type in @basic, do: &cast_basic(type, &1)
```

Comment:
What do you think about having different cast functions instead of `cast_basic`? It would pattern match only once then; and it wouldn't need the new `@basic` module attribute:

```elixir
defp caster(:float), do: &cast_float/1
defp caster(:boolean), do: &cast_boolean/1
defp caster(:binary_id), do: &cast_binary_id/1
defp caster(:id), do: &cast_integer/1
defp caster(:integer), do: &cast_integer/1
defp caster(:decimal), do: &cast_decimal/1
defp caster(:date), do: &cast_date/1
defp caster(:time), do: &maybe_truncate_usec(cast_time(&1))
defp caster(:time_usec), do: &maybe_pad_usec(cast_time(&1))
defp caster(:naive_datetime), do: &maybe_truncate_usec(cast_naive_datetime(&1))
defp caster(:naive_datetime_usec), do: &maybe_pad_usec(cast_naive_datetime(&1))
defp caster(:utc_datetime), do: &maybe_truncate_usec(cast_utc_datetime(&1))
defp caster(:utc_datetime_usec), do: &maybe_pad_usec(cast_utc_datetime(&1))
defp caster(type) when type in [:string, :binary, :map, :any], do: &cast_base(type, &1)

# rest of caster/1 functions

defp cast_float(term) when is_binary(term) do
  case Float.parse(term) do
    {float, ""} -> {:ok, float}
    _ -> :error
  end
end

defp cast_float(term) when is_integer(term), do: {:ok, :erlang.float(term)}
defp cast_float(term) when is_float(term), do: {:ok, term}
defp cast_float(_term), do: :error

defp cast_boolean(term) when term in ~w(true 1),  do: {:ok, true}
defp cast_boolean(term) when term in ~w(false 0), do: {:ok, false}
defp cast_boolean(_term), do: :error

defp cast_base(type, term) do
  if of_base_type?(type, term) do
    {:ok, term}
  else
    :error
  end
end
```

Reply:
In fact, this idea is not new in the codebase. It would be similar to how `equal_fun` is calculated:

```elixir
  defp equal_fun(nil), do: nil
  defp equal_fun(:decimal), do: &equal_decimal?/2
  defp equal_fun(t) when t in [:time, :time_usec], do: &equal_time?/2
  defp equal_fun(t) when t in [:utc_datetime, :utc_datetime_usec], do: &equal_utc_datetime?/2
  defp equal_fun(t) when t in [:naive_datetime, :naive_datetime_usec], do: &equal_naive_datetime?/2
  defp equal_fun(t) when t in @base, do: nil
```

The "catch-rest-base-type" clause could be changed then from 

```elixir
defp caster(type) when type in [:string, :binary, :map, :any], do: &cast_base(type, &1)
```

to

```elixir
defp caster(type) when type in @base, do: &cast_base(type, &1)
```

Reply:
The goal here is to avoid the type lookups, so we should do our best to avoid any type based dispatching.

Reply:
I think right now the type-based dispatching is split in two, first in the `caster(type) when type in @base` and then in the `cast_basic(:float, term)` or `cast_basic(:boolean, term)`. My idea was to merge the dispatching logic into the `caster` function. Would that make sense? :)